### PR TITLE
[TSDK 77] Outbox Support

### PR DESCRIPTION
**Pull request by mrashed-dev**
# Description
This PR brings support to the [new outbox beta](https://developer.nylas.com/docs/api#tag--Outbox).

# Usage
To send a new message through the outbox:

```java
NylasClient nylas = new NylasClient();
NylasAccount account = nylas.account("{ACCESS_TOKEN}");
Outbox outbox = account.outbox();

// Create a new outbox message object
OutboxMessage message = new OutboxMessage();
message.setSubject("With Love, From Nylas");
message.setTo(Arrays.asList(new NameEmail("My Nylas Friend", "swag@nylas.com")));
message.setBody("This email was sent using the Nylas email API. Visit https://nylas.com for details.");
message.setReplyTo(new NameEmail("Your Name", "you@example.com"));
message.setFrom(new NameEmail("Your Name", "you@example.com"));

// Set the outbox-specific parameters
Calendar calendar = Calendar.getInstance();
calendar.add(Calendar.DAY_OF_YEAR, 1);
Date tomorrow = calendar.getTime();
calendar.add(Calendar.DAY_OF_YEAR, 1);
Date dayAfter = calendar.getTime();

message.setSendAt(tomorrow);
message.setRetryLimitDatetime(dayAfter);

// Send the outbox message
OutboxJobStatus jobStatus = outbox.send(outboxMessage);

String outboxJobStatusId = jobStatus.getJobStatusId();
String outboxStatus = jobStatus.getStatus();
OutboxMessage sentMessage = jobStatus.getOriginalData();
```

You can also avoid creating an `OutboxMessage` type, you can just pass in a draft object to `send`:
```java
NylasClient nylas = new NylasClient();
NylasAccount account = nylas.account("{ACCESS_TOKEN}");
Outbox outbox = account.outbox();

// Create a new outbox message object
Draft draft = new Draft();
draft.setSubject("With Love, From Nylas");
draft.setTo(Arrays.asList(new NameEmail("My Nylas Friend", "swag@nylas.com")));
draft.setBody("This email was sent using the Nylas email API. Visit https://nylas.com for details.");
draft.setReplyTo(new NameEmail("Your Name", "you@example.com"));
draft.setFrom(new NameEmail("Your Name", "you@example.com"));

// Prepare the outbox-specific parameters
Calendar calendar = Calendar.getInstance();
calendar.add(Calendar.DAY_OF_YEAR, 1);
Date tomorrow = calendar.getTime();
calendar.add(Calendar.DAY_OF_YEAR, 1);
Date dayAfter = calendar.getTime();

// Send the outbox message
OutboxJobStatus jobStatus = outbox.send(draft, tomorrow, dayAfter);
```

To update an outbox message after sending it
```java
NylasClient nylas = new NylasClient();
NylasAccount account = nylas.account("{ACCESS_TOKEN}");
Outbox outbox = account.outbox();

// Create and send the message object
OutboxMessage message = new OutboxMessage();
message.setSubject("With Love, From Nylas");
message.setTo(Arrays.asList(new NameEmail("My Nylas Friend", "swag@nylas.com")));
message.setBody("This email was sent using the Nylas email API. Visit https://nylas.com for details.");
message.setReplyTo(new NameEmail("Your Name", "you@example.com"));
message.setFrom(new NameEmail("Your Name", "you@example.com"));

Calendar calendar = Calendar.getInstance();
calendar.add(Calendar.DAY_OF_YEAR, 1);
Date tomorrow = calendar.getTime();
calendar.add(Calendar.DAY_OF_YEAR, 1);
Date dayAfter = calendar.getTime();
message.setSendAt(tomorrow);
message.setRetryLimitDatetime(dayAfter);

OutboxJobStatus jobStatus = outbox.send(outboxMessage);

// Modify something in the job status
OutboxMessage sentMessage = jobStatus.getOriginalData();
calendar = Calendar.getInstance();
calendar.add(Calendar.HOUR_OF_DAY, 2);
Date newSendAt = calendar.getTime();

sentMessage.setSendAt(newSendAt);

// Update the outbox job
OutboxJobStatus updatedJobStatus = outbox.update(sentMessage, jobStatus.getJobStatusId());
```

To delete an outbox job
```java
NylasClient nylas = new NylasClient();
NylasAccount account = nylas.account("{ACCESS_TOKEN}");
Outbox outbox = account.outbox();

outbox.delete("JOB_STATUS_ID");
```

We also support SendGrid operations. To get the authentication and verification status
```java
NylasClient nylas = new NylasClient();
NylasAccount account = nylas.account("{ACCESS_TOKEN}");
Outbox outbox = account.outbox();

SendGridVerification verification = outbox.sendGridVerificationStatus();
verification.isDomainVerified();
verification.isSenderVerified();
```

To delete a SendGrid user:
```java
NylasClient nylas = new NylasClient();
NylasAccount account = nylas.account("{ACCESS_TOKEN}");
Outbox outbox = account.outbox();

outbox.deleteSendGridSubUser("email_address");
```

# License
I confirm that this contribution is made under the terms of the MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: src/main/java/com/nylas/NylasClient.java [hunk 1]
```diff
@@ -224,12 +261,19 @@ public NylasClient build() {
...
+		;
+	}
 	
 	static enum HttpMethod {
 		GET,
 		PUT,
 		POST,
 		DELETE,
+		PATCH,
 		;
```
Comment (mypile) [inline src/main/java/com/nylas/NylasClient.java:277]: nit: Do you need the `;` at the end here...? It looks weird from a formatting perspective.

File: src/main/java/com/nylas/OutboxJobStatus.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	}
+
+	@Override
+	public String toString() {
+		return "OutboxJobStatus [" +
+				"job_status_id='" + getJobStatusId() + '\'' +
+				"account_id='" + getAccountId() + '\'' +
+				", status='" + status + '\'' +
+				", original_data=" + original_data +
+				']';
```
Comment (mypile) [inline src/main/java/com/nylas/OutboxJobStatus.java:23]: nit: spacing here is weird. I think these lines can be in line with the first one?

File: src/main/java/com/nylas/OutboxMessage.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	@Override
+	public String toString() {
+		return "OutboxMessage [reply_to_message_id=" + reply_to_message_id + ", version=" + version + ", account_id="
+				+ getAccountId() + ", thread_id=" + thread_id + ", subject=" + subject + ", from=" + from + ", to=" + to
+				+ ", cc=" + cc + ", bcc=" + bcc + ", reply_to=" + reply_to + ", date=" + date + ", unread=" + unread
+				+ ", starred=" + starred + ", snippet=" + snippet + ", body=" + body + ", files=" + files + ", folder="
+				+ folder + ", labels=" + labels + ", tracking=" + tracking + ", metadata=" + metadata
+				+ ", send_at=" + send_at + ", retry_limit_datetime=" + retry_limit_datetime + "]";
```
Comment (mypile) [inline src/main/java/com/nylas/OutboxMessage.java:111]: nit: spacing here too.

File: src/main/java/com/nylas/SendGridVerifiedStatus.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	public boolean isSenderVerified() {
+		return sender_verified;
+	}
+
+	@Override
+	public String toString() {
+		return "SendGridVerifiedStatus [" +
+				"domain_verified=" + domain_verified +
+				", sender_verified=" + sender_verified +
+				']';
```
Comment (mypile) [inline src/main/java/com/nylas/SendGridVerifiedStatus.java:21]: nit: spacing here too.

File: src/main/java/com/nylas/NylasClient.java [hunk 5]
```diff
@@ -140,28 +161,44 @@ public NylasAccount account(String accessToken) {
...
 	}
 
 	<T> T executeRequestWithAuth(String authUser, HttpUrl.Builder url, HttpMethod method, RequestBody body,
 			Type resultType) throws IOException, RequestFailedException {
-		Request request = buildRequest(authUser, url, method, body);
+		return executeRequestWithAuth(authUser, url, method, body, resultType, AuthMethod.BASIC);
+	}
+
+	<T> T executeRequestWithAuth(String authUser, HttpUrl.Builder url, HttpMethod method, RequestBody body,
+								 Type resultType, AuthMethod authMethod) throws IOException, RequestFailedException {
```
Comment (mypile) [inline src/main/java/com/nylas/NylasClient.java:176]: nit: why is this line tabbed over so far...?

**Review by mypile — APPROVED** [review]
I can't speak to the functionality, but this LGTM outside of the nit comments.

File: src/main/java/com/nylas/NylasClient.java (hunk 1 above)
Comment (mrashed-dev) [reply src/main/java/com/nylas/NylasClient.java]: Thank you! Done ✅

File: src/main/java/com/nylas/OutboxJobStatus.java (hunk 2 above)
Comment (mrashed-dev) [reply src/main/java/com/nylas/OutboxJobStatus.java]: Thank you! Done ✅

File: src/main/java/com/nylas/OutboxMessage.java (hunk 3 above)
Comment (mrashed-dev) [reply src/main/java/com/nylas/OutboxMessage.java]: Thank you! Done ✅

File: src/main/java/com/nylas/SendGridVerifiedStatus.java (hunk 4 above)
Comment (mrashed-dev) [reply src/main/java/com/nylas/SendGridVerifiedStatus.java]: Thank you! Done ✅

File: src/main/java/com/nylas/NylasClient.java (hunk 5 above)
Comment (mrashed-dev) [reply src/main/java/com/nylas/NylasClient.java]: My Intellij settings were out of whack since the upgrade, fixed now! ✅

**Pull request merged**